src/pyodide/internal/workers.py

```diff
@@ -682,6 +719,138 @@ async def text(self) -> str:
         return await self.js_object.text()
 
 
+def serialize_to_python(obj: "JsProxy"):
+    """
+    Converts JS objects like Response, Request, Blob, etc. to equivalent Python objects defined in
+    this module and also other JS objects like Map, Set, etc. to equivalent Python stdlib objects.
+
+    This method is used for Workers RPC in Python to convert JavaScript objects to Python. As such
+    it does not support serializing all JS object types.
+    """
+
+    def convert(obj, convert, cache):
+        if not hasattr(obj, "constructor"):
+            # Assume that the object doesn't need conversion as it's not a JS object.
+            return obj
+
+        if obj.constructor.name == "Response":
+            return Response(obj)
+        elif obj.constructor.name == "FormData":
+            return FormData(obj)
+        elif obj.constructor.name == "Blob":
+            return Blob(obj)
+        elif obj.constructor.name == "File":
+            return File(obj)
+        elif obj.constructor.name == "Request":
+            return Request(obj)
+        elif obj.constructor.name == "Date":
+            # TODO: Pyodide should gain support for this, we should upstream this.
+            return datetime.datetime.fromtimestamp(obj.getTime() / 1000)
+        elif obj.constructor.name == "Error":
+            return Exception(obj.toString())
+        elif obj.constructor.name == "Number":
+            return obj.valueOf()
+
+        return obj
+
+    if not hasattr(obj, "constructor"):
+        return obj
+
+    result = obj.to_py(default_converter=convert)
+
+    if isinstance(result, JsProxy):
+        raise TypeError(
+            f"Couldn't convert object to Python type, got {obj.constructor.name}"
+        )
+
+    return result
+
+
+def _raise_on_disabled_type(value):
+    if isinstance(value, (tuple, bytearray)):
+        raise TypeError(f"{type(value)} cannot be sent over RPC.")
+
+    if _is_iterable(value):
+        for v in value:
+            _raise_on_disabled_type(v)
+
+
+def serialize_to_js(value) -> JsProxy:
+    """
+    Converts Python objects defined in this module (Response, Request, etc) and native Python types
+    like Map, Set, datetime to equivalent JavaScript types.
+
+    This method is used for Workers RPC in Python to convert Python objects to JavaScript. As such
+    it does not support serializing all Python object types.
+    """
+
+    def converter(obj, convert, cache):
```

Comment:
Also move this to a top level definition since it doesn't close over anything.

Reply:
Closing over variables isn't the only reason to create a nested function, I'd rather keep it in here if you don't mind. Grouping the code like this makes it easier to reason about IMO